src/Storages/MergeTree/ReplicatedMergeTreeCurrentlyRestoringFromBackup.cpp

```diff
@@ -0,0 +1,1126 @@
+            return;
+
+        LOG_INFO(log, "Checking that no parts exist in the table before restoring its data (reason: {})", magic_enum::enum_name(reason));
+
+        /// The order of checks is important:
+        /// an INSERT command first allocates a block number, then it creates a log entry, and finally it replicates the part into all replicas.
+        /// That's why it's more reliable to check first block numbers, then the replication queue and finally the current replica itself.
+
+        auto allocated_block_number = findAnyAllocatedBlockNumber(zookeeper, ignore_block_number_zk_paths);
+
+        auto part_name_in_queue = findAnyPartInReplicationQueue(zookeeper);
+
+        if (auto part_name_on_replica = findAnyPartOnReplica())
+            throwTableIsNotEmpty(reason, fmt::format("part {} exists", *part_name_on_replica));
+
+        if (part_name_in_queue)
+            throwTableIsNotEmpty(reason, fmt::format("part {} exists on some replicas", *part_name_in_queue));
+
+        if (auto restoring_part_name = findAnyRestoringPart(zookeeper))
+            throwTableIsNotEmpty(reason, fmt::format("concurrent RESTORE is creating part {}", *restoring_part_name));
+
+        if (allocated_block_number)
+            throwTableIsNotEmpty(reason, fmt::format("concurrent INSERT or ATTACH is using block number {} in partition {}", allocated_block_number->first, allocated_block_number->second));
+    }
+
+    /// Finds any part on the current replica.
+    std::optional<String> findAnyPartOnReplica() const
+    {
+        if (storage.getTotalActiveSizeInBytes() == 0)
+            return {};
+
+        auto parts_lock = storage.lockParts();
+
+        auto parts = storage.getDataPartsVectorForInternalUsage({MergeTreeDataPartState::Active, MergeTreeDataPartState::PreActive}, parts_lock);
+        if (parts.empty())
+            return {};
+
+        return parts.front()->name;
+    }
+
+    /// Finds any part in the replication queue.
+    std::optional<String> findAnyPartInReplicationQueue(const ZooKeeperWithFaultInjectionPtr & zookeeper) const
+    {
+        storage.queue.pullLogsToQueue(zookeeper, {}, ReplicatedMergeTreeQueue::CHECK_EMPTY);
+
+        auto queue_status = storage.queue.getStatus();
+        if (!queue_status.inserts_in_queue)
+            return {};
+
+        ReplicatedMergeTreeQueue::LogEntriesData log_entries;
+        storage.queue.getEntries(log_entries);
+
+        for (const auto & log_entry : log_entries)
+        {
+            if (log_entry.type == ReplicatedMergeTreeLogEntryData::GET_PART || log_entry.type == ReplicatedMergeTreeLogEntryData::ATTACH_PART)
+                return log_entry.new_part_name;
+        }
+
+        return {};
+    }
+
+    /// Finds any locked block number excluding ones we have just allocated.
+    std::optional<std::pair<Int64, String>> findAnyAllocatedBlockNumber(const ZooKeeperWithFaultInjectionPtr & zookeeper, const Strings & ignore_zk_paths) const
+    {
```

Comment:
See `ReplicatedMergeTreeMergePredicate::committing_blocks` and `BaseMergePredicate::getCommittingBlocks`, please reuse existing code when possible